vpr/src/pack/cluster.cpp

```diff
@@ -2527,33 +2726,105 @@ static void add_cluster_molecule_candidates_by_highfanout_connectivity(t_pb* cur
+ * Attraction groups can be very large, so we only add some randomly selected molecules for efficiency
+ * if the number of atoms in the group is greater than 500. Therefore, the molecules added to the candidates
+ * will vary each time you call this function.
  */
 static void add_cluster_molecule_candidates_by_attraction_group(t_pb* cur_pb,
                                                                 t_cluster_placement_stats* cluster_placement_stats_ptr,
                                                                 const std::multimap<AtomBlockId, t_pack_molecule*>& atom_molecules,
                                                                 AttractionInfo& attraction_groups,
-                                                                const int feasible_block_array_size) {
+                                                                const int feasible_block_array_size,
+                                                                ClusterBlockId clb_index,
+                                                                std::map<const t_model*, std::vector<t_logical_block_type_ptr>>& primitive_candidate_block_types) {
     auto& atom_ctx = g_vpr_ctx.atom();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+
+    auto cluster_type = cluster_ctx.clb_nlist.block_type(clb_index);
+
+    /*
+     * For each cluster, we want to explore the attraction group molecules as potential
+     * candidates for the cluster a limited number of times. This limit is imposed because
+     * if the cluster belongs to a very large attraction group, we could potentially search
+     * through its attraction group molecules for a very long time.
+     * Defining a number of times to search through the attraction groups (i.e. number of
+     * attraction group pulls) determines how many times we search through the cluster's attraction
+     * group molecules for candidate molecules.
+     */
+    int num_pulls = attraction_groups.get_att_group_pulls();
+    if (cur_pb->pb_stats->pulled_from_atom_groups < num_pulls) {
+        cur_pb->pb_stats->pulled_from_atom_groups++;
+    } else {
+        return;
+    }
 
     AttractGroupId grp_id = cur_pb->pb_stats->attraction_grp_id;
-    if (grp_id != AttractGroupId::INVALID()) {
-        AttractionGroup group = attraction_groups.get_attraction_group_info(grp_id);
+    if (grp_id == AttractGroupId::INVALID()) {
+        return;
+    }
 
-        for (AtomBlockId blk_id : group.group_atoms) {
-            if (atom_ctx.lookup.atom_clb(blk_id) == ClusterBlockId::INVALID()) {
-                auto rng = atom_molecules.equal_range(blk_id);
+    AttractionGroup& group = attraction_groups.get_attraction_group_info(grp_id);
+    int num_available_atoms = group.group_atoms.size();
+    if (num_available_atoms == 0) {
+        return;
+    }
+
+    if (num_available_atoms < 500) {
+        for (AtomBlockId atom_id : group.group_atoms) {
+            const auto& atom_model = atom_ctx.nlist.block_model(atom_id);
```

Comment:
Looks like duplicate code -- factor this and the code below into a helper function to avoid duplication (or more than one if necessary).